### PR TITLE
EOF: Implement `ext*calls`

### PR DESCRIPTION
**Pull request by rodiazet**
~Depends on #15620~ Merged.
~Depends on #15626~ Merged.

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1678]: Why?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (axic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2674]: Extcall?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (rodiazet) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1678]: regular `call` for this case was used only for rev before `staticcall` was introduced (before Byzantium). For EOF we have `extstaticcall` already.

**Comment by cameel** [comment]
Does this really depend on #15555? From a brief initial look it seems to me that they should be independent.

**Comment by rodiazet** [comment]
> Does this really depend on #15555? From a brief initial look it seems to me that they should be independent.

No. It does not.

File: test/libyul/yulSyntaxTests/eof/extcalls_invalid_in_legacy.yul
Comment (cameel) [inline test/libyul/yulSyntaxTests/eof/extcalls_invalid_in_legacy.yul]: This test should work the same for all EVM versions. It's enough that it's restricted to `legacy`.

I'd remove it from EOF-specific tests here too: `>=EOFv1` already ensures that we'll only run them on `prague` and later so restricting the EVM version is redundant.

File: libyul/AsmAnalysis.cpp
Comment (cameel) [inline libyul/AsmAnalysis.cpp]: Looks like an unrelated fix. Should be in a separate commit (or even PR).

BTW, this also shows that we're missing a proper test checking that the deprecated legacy opcodes are actually fobidden on EOF. We have `call_instruction_in_eof.yul` but that only covers `CALL`.

File: libyul/AsmAnalysis.cpp
Comment (cameel) [inline libyul/AsmAnalysis.cpp]: Is there any good reason to have an explicit check for specifically these three instructions? We did not do that for the ones we added earlier. Instead we just excluded them from dialect, so that they don't have corresponding builtins and their names still remain usable.

File: libevmasm/SemanticInformation.cpp
Comment (cameel) [inline libevmasm/SemanticInformation.cpp:174]: The only difference between these and `EXTCALL` seems to be number of args and returns. Might be better to unify them and make only that bit conditional.

File: libevmasm/SemanticInformation.cpp
Comment (cameel) [inline libevmasm/SemanticInformation.cpp]: The EOF versions only read memory. They don't accept a memory buffer to write the results like the legacy versions. And, unlike storage, the called contract cannot write to our memory.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (cameel) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: This is not necessary. We don't have `extcall`/`extdelegatecall`/`extstaticcall` as members on the `address` type.

`call`/`delegatecall`/`staticcall` are detected as the right `FunctionType::Kind` in analysis (`BareCall`/`BareDelegateCall`/`BareStaticCall`) and you already handle those in `appendBareCall()` so they will compile to the right opcode based on whether we're on EOF or not.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (cameel) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I'd suggest doing it the simple way. It's not really longer and has the advantage that it will actually show up when grepping the code for `extdelegatecall`:

```suggestion
		templ("call", eof ? "extdelegatecall" : "delegatecall");
```

Same for `extcall` and `extstaticcall`. And also in `appendBareCall()` below.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (cameel) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			let <success> := iszero(<call>(<address>, <pos>, <length> <?+value>, <value></+value>))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (cameel) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2872]: Please assert that in EOF `gas` is not set. Same in `appendExternalFunctionCall()`.

BTW, for this assert to be valid, we must also disallow setting `gas` on external calls via analysis. Please remember to do that later. Unless the plan is to just ignore `gas` on EOF for easier transition from legacy? But in that case we should still at least issue a warning in analysis.

File: test/libyul/yulSyntaxTests/eof/extcalls.yul
Comment (cameel) [inline test/libyul/yulSyntaxTests/eof/extcalls.yul]: Is the test passing in the wrong number of arguments on purpose? If so, why? Shouldn't it be testing that these builtins compile without errors?

Comment (cameel) [reply test/libyul/yulSyntaxTests/eof/extcalls.yul]: BTW, since testing `if`s and `revert` is not the purpose of the test, and we don't actually execute the code, you could make it much more concise:

```yul
pop(extcall(address(), 0, 0))
pop(extdelegatecall(address(), 0, 0, 10))
pop(extstaticcall(address(), 0, 0, 10))
```

Same applies to the other tests. Possibly using `sstore()` if `pop()` gets optimized out.

**Review by cameel — CHANGES_REQUESTED** [review]
There are a few things that need correcting: memory side-effects, non-existent `ext` members on `address`, probably unnecessary check for new opcodes outside of EOF. We also have no analysis checks against passing `gas` into an external call.

Also, tests are failing.

**Comment by cameel** [comment]
I think it would be good to enable a few related semantic tests already in this PR. Some external calls and the low-level calls on `address`. Otherwise we have zero coverage for the codegen changes here and you will have to deal with them all at once when we enable all semantic tests.

**Comment by cameel** [comment]
Also, a question about the [EOF spec](https://github.com/ipsilon/eof/blob/main/spec/eof.md). It says:

> **NOTE**: The replacement instructions `EXT*CALL` continue being treated as **undefined** in legacy code.

Does it simply mean that the opcodes are treated as `INVALID` on legacy? Isn't this the case for all newly introduced opcodes? I have an impression that there might be more to this than that because otherwise why call out specifically `EXT*CALL`?

**Comment by rodiazet** [comment]
Test are failing because they depend on other PR with stack height calculation. I can regenerate them but I would rather wait until it's merged.

File: test/libyul/yulSyntaxTests/eof/extcalls_invalid_in_legacy.yul
Comment (rodiazet) [reply test/libyul/yulSyntaxTests/eof/extcalls_invalid_in_legacy.yul]: done.

File: libyul/AsmAnalysis.cpp
Comment (rodiazet) [reply libyul/AsmAnalysis.cpp]: Because these three opcodes are available in yul (EOF) as regular instructions. Other instructions I added are buildin functions or are unavailable in yul. If we don't add [the sanity check](https://github.com/ethereum/solidity/blob/develop/libyul/AsmAnalysis.cpp#L804) fails

File: libyul/AsmAnalysis.cpp
Comment (rodiazet) [reply libyul/AsmAnalysis.cpp]: Will move to separate. We have https://github.com/ethereum/solidity/pull/15559/files#diff-30c103e3872c974dedcee26006d5ffad9eca618467cc82a4f7f1aed465c16fae for legacy calls.

File: libevmasm/SemanticInformation.cpp
Comment (rodiazet) [reply libevmasm/SemanticInformation.cpp:174]: Right. Fixed.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (rodiazet) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2872]: In this PR?

File: test/libyul/yulSyntaxTests/eof/extcalls.yul
Comment (rodiazet) [reply test/libyul/yulSyntaxTests/eof/extcalls.yul]: done

**Comment by rodiazet** [comment]
> I think it would be good to enable a few related semantic tests already in this PR. Some external calls and the low-level calls on `address`. Otherwise we have zero coverage for the codegen changes here and you will have to deal with them all at once when we enable all semantic tests.

I have already a version with all semantic tests passing. It required three small fixes but in this case I would not bother now. Moreover we need to introduce Osaka to Solidity to run semantic tests with the newest evmone now.

**Comment by gumb0** [comment]
> Also, a question about the [EOF spec](https://github.com/ipsilon/eof/blob/main/spec/eof.md). It says:
> 
> > **NOTE**: The replacement instructions `EXT*CALL` continue being treated as **undefined** in legacy code.
> 
> Does it simply mean that the opcodes are treated as `INVALID` on legacy? Isn't this the case for all newly introduced opcodes? I have an impression that there might be more to this than that because otherwise why call out specifically `EXT*CALL`?

It's the same as with all other newly introduced opcodes.

Original plan was to enable them also for legacy, then it changed, and this explicit note added to point out the change.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (cameel) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2872]: I think that the asserts should be added already here as a quick sanity check. The analysis error or warning can be done separately.

**Comment by cameel** [comment]
> Test are failing because they depend on other PR with stack height calculation. I can regenerate them but I would rather wait until it's merged.

Ah, ok. I guess that may have been the reason for making it a dependency?

In any case that PR is merged now.

File: libyul/AsmAnalysis.cpp
Comment (cameel) [reply libyul/AsmAnalysis.cpp]: > We have https://github.com/ethereum/solidity/pull/15559/files#diff-30c103e3872c974dedcee26006d5ffad9eca618467cc82a4f7f1aed465c16fae for legacy calls.


The link does not point me at any specific file... I guess you meant that we already have a test covering all legacy-only opcodes? Which one?

**Comment by rodiazet** [comment]
One dependency added

File: libyul/AsmAnalysis.cpp
Comment (rodiazet) [reply libyul/AsmAnalysis.cpp]: `test/libyul/yulSyntaxTests/eof/legacy_calls_in_eof.yul` now moved to https://github.com/ethereum/solidity/pull/15620

File: libevmasm/SemanticInformation.cpp
Comment (rodiazet) [reply libevmasm/SemanticInformation.cpp]: forgot about it. Fixed now

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (rodiazet) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: ok

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (rodiazet) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: changed

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (rodiazet) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2872]: Asserts added. I meant the warning.

File: libyul/AsmAnalysis.cpp
Comment (rodiazet) [reply libyul/AsmAnalysis.cpp]: I mean all added opcodes were till now buildin or are disabled in dialect.

Comment (cameel) [reply libyul/AsmAnalysis.cpp]: Ah, right. I'm still confusing this bit with the code responsible for making sure instructions are available or not on EOF/legacy. But this only kicks in when we would get an error anyway and just replaces the generic "function not found" with a more specific message explaining why it was not available.

So this is correct, it's the other way around - the dialect builtins should report these specialized messages too if we wanted to be fully consistent.

Comment (cameel) [reply libyul/AsmAnalysis.cpp]: But while thinking about it I realized we're missing a check elsewhere and a very important one - we're missing the special-casing for the `EXT*CALL` instructions on legacy. They should be marked as non-reserved in `EVMDialect::createReservedIdentifiers()`. Without it, this is breaking for legacy Yul.

Comment (cameel) [reply libyul/AsmAnalysis.cpp]: And it's actually not just this PR, we're missing such exceptions for nearly all the instructions we added so far. We did that only for `auxdataloadn`, but then we forgot `dataloadn`, which gets added automatically as all `Instruction`s are.

This thing is just too error-prone and we keep realizing we're missing something there (https://github.com/ethereum/solidity/pull/15550#discussion_r1859401044). I think that instead of testing it ad-hoc for every new opcode/builtin, we should create a few comprehensive test cases that cover all of them and just keep them updated with new PRs.

We basically need 4 tests. Each one should cover all EOF-only opcodes and all EOF-only builtins and check that:
1. The name is **not** reserved on legacy (can declare variables and functions with that name). Covers `createReservedIdentifiers()`.
2. The name is reserved on EOF.
3. The name is **not** defined on legacy (calling that name will not execute a Yul builtin). Covers `createBuiltins()`.
    - If it works on EOF, we get a specialized error saying it's only available on EOF (rather than a generic error about calling undefined function). Covers `validateInstructions()` we have here.
4. It either works on EOF or is not defined (depending on the case).

This can of course be done in a separate PR not to hold this one up. In this one just make sure that these things work correctly for `EXT*CALL` instructions.

Comment (cameel) [inline libyul/AsmAnalysis.cpp]: Since we're keeping this bit after all, I'll add one more suggestion:

```suggestion
				"The \"{}\" instruction is only available on EOF.",
				fmt::arg("instruction", boost::to_lower_copy(instructionInfo(_instr, m_evmVersion).name))
```

This is more concise and also the `kind` makes this confusing to me every time I look at this message because it's not kind of something, it's just a part of the message.

Also I'd refer to EOF explicitly. "Legacy" can be many things and we're generally overusing it in the compiler. Even in this context there is one other thing it could refer to - the older EVMs we deprecated support for. Saying "EOF" removes the ambiguity. Since this code is here solely to give a more specific error message, clarity is the whole point :)

Would be nice to change the one below to match too.

File: libyul/AsmAnalysis.cpp
Comment (cameel) [reply libyul/AsmAnalysis.cpp]: Ok, that covers this specific case, but I still think we need one that just covers all the opcodes, because we keep running into bugs like this.

Along the same lines as what I just suggested in https://github.com/ethereum/solidity/pull/15559#discussion_r1873671693, but for legacy-only opcodes.

Comment (cameel) [reply libyul/AsmAnalysis.cpp]: Just to be clear, does not have to be done in this PR. I guess we can continue this discussion in #15620.

**Review by cameel — COMMENTED** [review]
The main remaining thing is https://github.com/ethereum/solidity/pull/15559#discussion_r1873660414, which I missed previously. I.e. `extcall` opcodes must not be reserved identifiers on legacy + a test confirming it works correctly by defining a function/variable with that name.

The same thing needs to be fixed for other opcodes, but that should be done separately, since that's already broken on `develop`.

> I have already a version with all semantic tests passing. It required three small fixes but in this case I would not bother now. Moreover we need to introduce Osaka to Solidity to run semantic tests with the newest evmone now.

Well, if you prefer we can merge without it, but I'd still recommend enabling at least one just to confirm that late changes in the PR did not break anything.

File: libyul/AsmAnalysis.cpp
Comment (rodiazet) [reply libyul/AsmAnalysis.cpp]: Addressed. I created these tests covering `ext*call` in this PR. They should be extended to all added EOF instructions

**Comment by rodiazet** [comment]
> The main remaining thing is [#15559 (comment)](https://github.com/ethereum/solidity/pull/15559#discussion_r1873660414), which I missed previously. I.e. `extcall` opcodes must not be reserved identifiers on legacy + a test confirming it works correctly by defining a function/variable with that name.
> 
> The same thing needs to be fixed for other opcodes, but that should be done separately, since that's already broken on `develop`.
> 
> > I have already a version with all semantic tests passing. It required three small fixes but in this case I would not bother now. Moreover we need to introduce Osaka to Solidity to run semantic tests with the newest evmone now.
> 
> Well, if you prefer we can merge without it, but I'd still recommend enabling at least one just to confirm that late changes in the PR did not break anything.

I enabled one but to be able to run any semantic test we need to first merge this https://github.com/ethereum/solidity/pull/15626 or https://github.com/ethereum/solidity/pull/15606

**Review by cameel — APPROVED** [review]

**Pull request merged**
